packages/flutter_tools/lib/src/base/net.dart

```diff
@@ -104,15 +133,81 @@ Future<List<int>> _attempt(Uri url, { bool onlyHeaders = false }) async {
     }
     // 5xx errors are server errors and we can try again
     printTrace('Download error: ${response.statusCode} ${response.reasonPhrase}');
-    return null;
+    return false;
   }
   printTrace('Received response from server, collecting bytes...');
   try {
-    final BytesBuilder responseBody = BytesBuilder(copy: false);
-    await response.forEach(responseBody.add);
-    return responseBody.takeBytes();
+    assert(destSink != null);
+    await response.forEach(destSink.add);
+    return true;
   } on IOException catch (error) {
     printTrace('Download error: $error');
-    return null;
+    return false;
+  } finally {
+    await destSink?.flush();
+    await destSink?.close();
+  }
+}
+
+/// An IOSink that collects whatever is written to it.
+class _MemoryIOSink implements IOSink {
+  @override
+  Encoding encoding = utf8;
+
+  final BytesBuilder writes = BytesBuilder(copy: false);
+
+  @override
+  void add(List<int> data) {
+    writes.add(data);
+  }
+
+  @override
+  Future<void> addStream(Stream<List<int>> stream) {
+    final Completer<void> completer = Completer<void>();
+    stream.listen((List<int> data) {
```

Comment:
Could this be `stream.listen(add).onDone(completer.complete)`?

Reply:
Done